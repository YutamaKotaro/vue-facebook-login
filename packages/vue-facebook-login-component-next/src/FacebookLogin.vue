<template>
  <!-- https://v3.vuejs.org/guide/migration/v-bind.html#frontmatter-title -->
  <!-- https://stackoverflow.com/a/50892881/4106263 -->
  <scope v-bind="$props" v-model="model" v-slot="slotProps">
    <Button v-bind="{ ...buttonProps, ...$attrs }" :scope="slotProps">
      <template
        v-for="(slot, slotName) of $slots"
        v-slot:[slotName]="slotProps"
      >
        <slot v-bind="slotProps" :name="slotName" />
      </template>
    </Button>
  </scope>
</template>

<script>
import Scope from '@/Scope'
import Button from '@/Button.vue'
import { getComponentProps } from '@/utils'

export default {
  name: 'facebook-login',
  components: { Scope, Button },
  props: { ...Scope.props, ...Button.props },
  computed: {
    model: {
      get() {
        return this.value
      },
      set(value) {
        this.$emit('input', value)
      },
    },
    buttonProps() {
      return getComponentProps(Button, this.$props)
    },
  },
}
</script>
