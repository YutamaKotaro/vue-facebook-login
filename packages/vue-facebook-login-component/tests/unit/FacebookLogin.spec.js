import { mount } from '@vue/test-utils'
import Scope from '@/components/Scope'
import FacebookLogin from '@/components/FacebookLogin'
import { flush, mockSdk } from './test-utils'
import { Sdk } from '@/Sdk'

jest.mock('@/Sdk', () => {
  const { Sdk } = jest.requireActual('@/Sdk')
  const { mockSdk } = jest.requireActual('./test-utils')
  mockSdk(Sdk)
  return {
    __esModule: true,
    ...jest.requireActual('@/Sdk'),
    Sdk,
  }
})

const consoleError = console.error
const commonProps = { appId: '966242223397117' }

beforeEach(() => {
  console.error = consoleError
  mockSdk(Sdk)
})

jest.useFakeTimers()

describe('FacebookLogin', () => {
  test('initial render (disabled)', async () => {
    console.error = jest.fn()
    const wrapper = mount(FacebookLogin)
    await flush()
    expect(wrapper.attributes('disabled')).toBe('disabled')
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('initial render (disconnected)', async () => {
    const wrapper = mount(FacebookLogin, { propsData: commonProps })
    await flush()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('initial render (connected)', async () => {
    Sdk.getLoginStatus.mockResolvedValueOnce({ status: 'connected' })
    const wrapper = mount(FacebookLogin, {
      propsData: commonProps,
    })
    await flush()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('login and click event', async () => {
    const onClick = jest.fn()
    const wrapper = mount(FacebookLogin, {
      propsData: commonProps,
      listeners: { click: onClick },
    })
    await flush()
    wrapper.trigger('click')
    await flush()
    expect(onClick).toHaveBeenCalled()
    expect(wrapper.emitted().input.slice(-1)[0]).toEqual([
      {
        idle: true,
        error: null,
        hasError: false,
        enabled: true,
        working: false,
        disabled: false,
        connected: true,
        disconnected: false,
      },
    ])
  })

  test('props', async () => {
    const propsData = {
      appId: '966242223397117',
      version: 'mock-version',
      value: { connected: false },
      options: { option1: 1, option2: 2, option3: 3 },
      loginOptions: { loginOption1: 1, loginOption2: 2, loginOption3: 3 },
      scope: {},
      logoClass: 'logo-class',
      logoStyle: { logoStyle1: 1, logoStyle2: 2, logoStyle3: 3 },
      textClass: 'text-class',
      textStyle: { textStyle1: 1, textStyle2: 2, textStyle3: 3 },
      loaderClass: 'loader-class',
      loaderStyle: { loaderStyle1: 1, loaderStyle2: 2, loaderStyle3: 3 },
      transition: [
        'background-color 0.15s ease-in-out',
        'padding 0.15s ease-in-out',
      ],
      useAltLogo: true,
      asyncDelay: 500,
      sdkLocale: 'en_US',
    }
    const wrapper = mount(FacebookLogin, { propsData })
    expect(wrapper.vm.$props).toEqual(propsData)
  })

  test('slots', async () => {
    const wrapper = mount(FacebookLogin, {
      propsData: { appId: '966242223397117' },
      slots: {
        logo: '<div class="logo"></div>',
        after: '<div class="after"></div>',
        before: '<div class="before"></div>',
        login: '<div class="login"></div>',
        error: '<div class="error"></div>',
        working: '<div class="working"></div>',
      },
    })
    await flush()
    const scopeInstance = wrapper.get(Scope).vm
    scopeInstance.error = new Error()
    scopeInstance.$forceUpdate()
    await scopeInstance.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('scoped-slots', async () => {
    const wrapper = mount(FacebookLogin, {
      propsData: commonProps,
      scopedSlots: {
        logo: '<div class="logo" slot-scope="scope">{{scope}}</div>',
        after: '<div class="after" slot-scope="scope">{{scope}}</div>',
        before: '<div class="before" slot-scope="scope">{{scope}}</div>',
        login: '<div class="login" slot-scope="scope">{{scope}}</div>',
        error: '<div class="error" slot-scope="scope">{{scope}}</div>',
        working: '<div class="working" slot-scope="scope">{{scope}}</div>',
      },
    })
    await flush()
    const scopeInstance = wrapper.get(Scope).vm
    scopeInstance.error = new Error()
    scopeInstance.$forceUpdate()
    await scopeInstance.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('logout slot', async () => {
    Sdk.getLoginStatus.mockResolvedValueOnce({ status: 'connected' })
    const wrapper = mount(FacebookLogin, {
      propsData: commonProps,
      slots: { logout: '<div class="logout"></div>' },
    })
    await flush()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('logout scoped-slot', async () => {
    Sdk.getLoginStatus.mockResolvedValueOnce({ status: 'connected' })
    const wrapper = mount(FacebookLogin, {
      propsData: commonProps,
      scopedSlots: {
        logout: '<div class="logout" slot-scope="scope">{{scope}}</div>',
      },
    })
    await flush()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
