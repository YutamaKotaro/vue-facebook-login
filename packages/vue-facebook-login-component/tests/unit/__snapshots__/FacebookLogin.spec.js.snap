// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacebookLogin initial render (connected) 1`] = `
<button class="v-facebook-login" style="transition: background-color 0.15s ease-in-out;">
  <span> Logout </span>
</button>
`;

exports[`FacebookLogin initial render (disabled) 1`] = `
<button class="v-facebook-login" disabled="disabled" style="transition: background-color 0.15s ease-in-out;">
  <svg class="logo" fill="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
    <path d="M22.675 0h-21.35c-.732 0-1.325.593-1.325 1.325v21.351c0 .731.593 1.324 1.325 1.324h11.495v-9.294h-3.128v-3.622h3.128v-2.671c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.795.143v3.24l-1.918.001c-1.504 0-1.795.715-1.795 1.763v2.313h3.587l-.467 3.622h-3.12v9.293h6.116c.73 0 1.323-.593 1.323-1.325v-21.35c0-.732-.593-1.325-1.325-1.325z" />
  </svg>
  <span>Continue with Facebook </span>
</button>
`;

exports[`FacebookLogin initial render (disconnected) 1`] = `
<button class="v-facebook-login" style="transition: background-color 0.15s ease-in-out;">
  <svg class="logo" fill="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
    <path d="M22.675 0h-21.35c-.732 0-1.325.593-1.325 1.325v21.351c0 .731.593 1.324 1.325 1.324h11.495v-9.294h-3.128v-3.622h3.128v-2.671c0-3.1 1.893-4.788 4.659-4.788 1.325 0 2.463.099 2.795.143v3.24l-1.918.001c-1.504 0-1.795.715-1.795 1.763v2.313h3.587l-.467 3.622h-3.12v9.293h6.116c.73 0 1.323-.593 1.323-1.325v-21.35c0-.732-.593-1.325-1.325-1.325z" />
  </svg>
  <span>Continue with Facebook </span>
</button>
`;

exports[`FacebookLogin logout scoped-slot 1`] = `
<button class="v-facebook-login" style="transition: background-color 0.15s ease-in-out;">
  <span>
    <div class="logout">{
      "idle": true,
      "error": null,
      "hasError": false,
      "working": false,
      "enabled": true,
      "disabled": false,
      "connected": true,
      "disconnected": false
      }</div>
  </span>
</button>
`;

exports[`FacebookLogin logout slot 1`] = `
<button class="v-facebook-login" style="transition: background-color 0.15s ease-in-out;">
  <span>
    <div class="logout"></div>
  </span>
</button>
`;

exports[`FacebookLogin scoped-slots 1`] = `
<button class="v-facebook-login" disabled="disabled" style="transition: background-color 0.15s ease-in-out;">
  <div class="before">{
    "idle": false,
    "error": {},
    "hasError": true,
    "working": false,
    "enabled": false,
    "disabled": true,
    "connected": false,
    "disconnected": true
    }</div>
  <span>
    <div class="error">{
      "idle": false,
      "error": {},
      "hasError": true,
      "working": false,
      "enabled": false,
      "disabled": true,
      "connected": false,
      "disconnected": true
      }</div>
  </span>
  <div class="after">{
    "idle": false,
    "error": {},
    "hasError": true,
    "working": false,
    "enabled": false,
    "disabled": true,
    "connected": false,
    "disconnected": true
    }</div>
</button>
`;

exports[`FacebookLogin slots 1`] = `
<button class="v-facebook-login" disabled="disabled" style="transition: background-color 0.15s ease-in-out;">
  <div class="before"></div>
  <span>
    <div class="error"></div>
  </span>
  <div class="after"></div>
</button>
`;
